sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,772 @@
+   * cause any side-effects and the hardware is in a valid and recoverable
+   * state.
+   */
+  kDifUsbdevRxPacketGetInfoResultBadArg = 2,
+  /**
+   * Indicates that there is already a read buffer operation in progress.
+   */
+  kDifUsbdevRxPacketGetInfoResultBusy,
+  /**
+   * Indicates that RX FIFO is empty.
+   */
+  kDifUsbdevRxPacketGetInfoResultNoPackets,
+} dif_usbdev_rx_packet_get_info_result_t;
+
+/**
+ * Get information about the packet at the front of RX FIFO.
+ *
+ * The USB device has a small FIFO (RX FIFO) that stores received packets until
+ * the software has a chance to process them. It is the responsibility of the
+ * software to ensure that the RX FIFO is never full. If the host tries to send
+ * a packet when the RX FIFO is full, the USB device will respond with a NAK.
+ * While this will typically cause the host to retry transmission for regular
+ * data packets, there are transactions in the USB protocol during which the USB
+ * device is not allowed to send a NAK. Thus, the software must read received
+ * packets as soon as possible.
+ *
+ * Reading received packets involves two main steps:
+ * - Calling this function, i.e. `dif_usbdev_rx_packet_get_info`, and
+ * - Calling `dif_usbdev_rx_packet_read_bytes` until the entire packet payload
+ * is read or calling `dif_usbdev_rx_packet_discard_bytes`.
+ *
+ * In order to read an incoming packet, clients should first call this function
+ * to get information about the packet. In addition to populating a
+ * `dif_usbdev_rx_packet_info_t` struct, this also starts an internal read
+ * buffer operation. The state of this operation, i.e. the buffer to read from,
+ * number of bytes remaining and the offset to start reading from, is tracked by
+ * this library, therefore clients do not need to perform any additional state
+ * tracking. Then, clients should call `dif_usbdev_rx_packet_read_bytes` one or
+ * more times (depending on the sizes of their internal buffers) until the
+ * entire packet payload is read. Once the entire packet is read, the read
+ * buffer operation ends and the buffer that holds the packet payload is
+ * returned to the free buffer pool. If the clients want to ignore the payload
+ * of a packet, e.g. for an unsupported or a zero-length packet, they can call
+ * `dif_usbdev_rx_packet_discard_bytes` to immediately return the buffer that
+ * holds the packet payload to the free buffer pool and end the read buffer
+ * operation.
+ *
+ * @param usbdev A USB device.
+ * @return The result of the operation.
+ */
+dif_usbdev_rx_packet_get_info_result_t dif_usbdev_rx_packet_get_info(
+    dif_usbdev_t *usbdev, dif_usbdev_rx_packet_info_t *packet);
+
+/**
+ * Return codes for `dif_usbdev_rx_packet_read_bytes`.
+ */
+typedef enum dif_usbdev_rx_packet_read_bytes_result {
+  /**
+   * Indicates that the call succeeded and the entire packet payload is read.
+   */
+  kDifUsbdevRxPacketReadBytesResultOK = 0,
```

Comment:
Rather than doing `= 0`, Maybe these should reference the constants in the "main" error enum?

Reply:
I agree.